src/repo.rs

```diff
@@ -490,43 +428,175 @@ impl<T: NetworkAdapter + 'static> Repo<T> {
+        S: Sink<NetworkMessage, Error = SendErr> + Stream<Item = Result<NetworkMessage, RecvErr>>,
+    {
+        let (mut sink, stream) = stream.split();
+        let to_send = self.outgoing(collection).map(|m| Ok::<_, SendErr>(m));
+        let sending = to_send.forward(&mut sink).fuse();
+
+        let recv_msg = self.collection_sender.clone();
+        let recving = stream
+            .map_err(error::RunConnection::<SendErr, RecvErr>::Recv)
+            .try_for_each({
+                move |m| {
+                    futures::future::ready(
+                        recv_msg
+                            .try_send((collection, CollectionEvent::Network(m.into())))
+                            .map_err(|_| error::RunConnection::<SendErr, RecvErr>::RecvCollection),
+                    )
+                }
+            })
+            .fuse();
+
+        let finished = futures::future::select(sending, recving).await;
+        match finished {
+            futures::future::Either::Left((result, _recving)) => match result {
+                Ok(_) => {
+                    tracing::trace!("we're closing the connection");
+                    Ok(())
+                }
+                Err(e) => {
+                    tracing::trace!(err=?e, "error while sending");
+                    Err(error::RunConnection::Send(e))
+                }
+            },
+            futures::future::Either::Right((result, _sending)) => {
+                if let Err(e) = sink.close().await {
+                    tracing::trace!(err=?e, "error while closing sink");
+                }
+                match result {
+                    Ok(_) => {
+                        tracing::trace!("the remote closed the connection");
+                        Ok(())
+                    }
+                    Err(e) => {
+                        tracing::trace!(err=?e, "error while receiving");
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    fn outgoing(&self, collection: CollectionId) -> impl Stream<Item = NetworkMessage> {
+        OutgoingMessages {
+            outboxes: self.outboxes.clone(),
+            id: collection,
+        }
+    }
+
+    pub fn shutdown(&self) {
+        self.control_sender.send(ControlEvent::Shutdown).unwrap();
+    }
+}
+
+struct OutgoingMessages {
+    outboxes: Arc<Mutex<HashMap<CollectionId, Outbox>>>,
+    id: CollectionId,
+}
+
+impl Stream for OutgoingMessages {
+    type Item = NetworkMessage;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut outboxes = self.outboxes.lock();
+        if let Some(Outbox { waker, messages }) = outboxes.get_mut(&self.id) {
+            if let Some(msg) = messages.pop() {
+                Poll::Ready(Some(msg))
+            } else {
+                *waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+        } else {
+            outboxes.insert(
```

Comment:
This still would leave the end-user multiplexing multiple peers into a single stream/outbox per collection right? I thought a goal was to facilitate a "one stream per peer" setup?

How about we make these concepts a "one per `RepoId`" with one?

Reply:
Yeah correct, I was aiming to just get a sketch of what the API looks like so there are a few holes. There should really be some kind of "connection" abstraction in the `Repo` with an outbox per connection.